packages/runtime/container-runtime/src/summarizer.ts

```diff
@@ -100,17 +106,107 @@ export interface ISummarizerEvents extends IEvent {
+     * Normally, the parent client would realize it is disconnected first and call stop
+     * giving a "parentNotConnected" stop reason. If the summarizer client attempts to
+     * generate a summary and realizes at that moment that the parent is not elected,
+     * only then will it stop itself with this message.
+     */
+    | "parentNoLongerSummarizer"
+    /** Parent client reported that it is no longer connected. */
+    | "parentNotConnected"
+    /**
+     * Parent client reported that it is no longer elected the summarizer.
+     * This is the normal flow; a disconnect will always trigger the parent
+     * client to no longer be elected as responsible for summaries. Then it
+     * tries to stop its spawned summarizer client.
+     */
+    | "parentShouldNotSummarize"
+    /** Parent client reported that it is disposed. */
+    | "disposed";
 export interface ISummarizer
     extends IEventProvider<ISummarizerEvents>, IFluidRouter, IFluidRunnable, IFluidLoadable {
     /**
      * Returns a promise that will be resolved with the next Summarizer after context reload
      */
     setSummarizer(): Promise<Summarizer>;
-    stop(reason?: string): void;
+    stop(reason?: SummarizerStopReason): void;
     run(onBehalfOf: string): Promise<void>;
     updateOnBehalfOf(onBehalfOf: string): void;
 }
 
+type SummarizeReason =
+    /**
+     * Attempt to summarize after idle timeout has elapsed.
+     * Idle timer restarts whenever an op is received. So this
+     * triggers only after some amount of time has passed with
+     * no ops being received.
+     */
+    | "idle"
+    /**
+     * Attempt to summarize after a maximum time since last
+     * successful summary has passed. This measures time since
+     * last summary ack op was processed.
+     */
+    | "maxTime"
+    /**
+     * Attempt to summarize after a maximum number of ops have
+     * passed since the last successful summary. This compares
+     * op sequence numbers with the reference sequence number
+     * of the summarize op corresponding to the last summary
+     * ack op.
+     */
+    | "maxOps"
+    /**
+     * Special case to generate a summary immediately after starting.
+     * @deprecated - only used in context reload hot swap, which was removed
+     */
+    | "immediate"
+    /**
+     * Special case to generate a summary in response to a Save op.
+     * @deprecated - do not use save ops
+     */
+    | `;${string}: ${string}`
+    /**
+     * Special case to attempt to summarize one last time before the
+     * summarizer client closes itself. This is to prevent cases where
+     * the summarizer client never gets a chance to summarize, because
+     * there are too many outstanding ops and/or parent client cannot
+     * stay connected long enough for summarizer client to catch up.
+     */
+    | "lastSummary"
+    /** Previous summary attempt failed, and we are retrying. */
+    | `retry${1 | 2}`;
```

Comment:
Note: Previously, we would reuse the same reason message that originally triggered the summary for the retries as well. Now it will replace it with this code; for example:
1. "maxOps"
2. "retry1"
3. "retry2"